File: py/portfolio.py
from money import Money
from bank import Bank


class Portfolio:
    def __init__(self) -> None:
        self._moneys = []

    def add(self, *moneys: list) -> None:
        self._moneys.extend(moneys)

    def evaluate(self, bank: Bank, currency: str) -> Money:
        total = Money(0, currency)
        failures = []

        for money in self._moneys:
            converted_currency, error = bank.convert(money, currency)
            if not error:
                total += converted_currency
            else:
                failures.append(error)

        if failures:
            raise Exception(f'Missing exchange rates: [{", ".join(failures)}]')

        return total

File: js/portfolio.js
import Money from './money.js';

class Portfolio {
  constructor() {
    this.moneys = [];
  }

  add(...moneys) {
    this.moneys = this.moneys.concat(moneys);
  }

  evaluate(bank, currency) {
    const failures = [];

    const total = this.moneys.reduce((sum, money) => {
      try {
        const convertedMoney = bank.convert(money, currency);
        return sum.add(convertedMoney);
      } catch (error) {
        failures.push(error.message);
        return sum;
      }
    }, new Money(0, currency));

    if (failures.length) {
      throw new Error(`Missing exchange rates: [${failures.join(', ')}]`);
    }

    return total;
  }
}

export default Portfolio;
